### PR TITLE
iss-1416

### PR DESCRIPTION
**Pull request by emfusion**
- fix get_bgp_neighbors_detail() when using  multi-agent mode
- fix issue where multihop var was wrongly interpreted as it was comparing 0 with ‘0’
- Add IPv6 prefixes to the advertised and received prefix count
- Add more test for get_bgp_neighbors_detail

### CODE REVIEW

File: napalm/eos/eos.py [hunk 1]
```diff
@@ -1682,13 +1682,26 @@ def _parse_per_peer_bgp_detail(peer_output):
                 "configured_keepalive",
                 "advertised_prefix_count",
                 "received_prefix_count",
+                "advertised_ipv6_prefix_count",
+                "received_ipv6_prefix_count",
             ]
 
             peer_details = []
 
+            # determine if in multi-agent mode to get correct extrator_type
```
Comment (fdomain) [inline napalm/eos/eos.py:1691]: typo: extractor

File: napalm/eos/eos.py [hunk 2]
```diff
@@ -1740,6 +1753,12 @@ def _parse_per_peer_bgp_detail(peer_output):
                 item["local_address"] = napalm.base.helpers.convert(
                     napalm.base.helpers.ip, item["local_address"]
                 )
+                # Get all the advertised prefixes
+                item["advertised_prefix_count"] += item["advertised_ipv6_prefix_count"]
+                item["received_prefix_count"] += item["received_ipv6_prefix_count"]
+                # Remove the ipv6_prefix for comformity with test_model
```
Comment (fdomain) [inline napalm/eos/eos.py:1759]: typo: conformity

File: napalm/eos/eos.py [hunk 3]
```diff
@@ -1740,6 +1753,12 @@ def _parse_per_peer_bgp_detail(peer_output):
                 item["local_address"] = napalm.base.helpers.convert(
                     napalm.base.helpers.ip, item["local_address"]
                 )
+                # Get all the advertised prefixes
+                item["advertised_prefix_count"] += item["advertised_ipv6_prefix_count"]
+                item["received_prefix_count"] += item["received_ipv6_prefix_count"]
```
Comment (fdomain) [inline napalm/eos/eos.py:1758]: Why do you mix ipv4 and ipv6 prefixes in the same counter ?
Shouldn't be ipv6 prefixes counter added to its own counter ?

File: napalm/eos/eos.py [hunk 4]
```diff
@@ -1774,7 +1793,7 @@ def _append(bgp_dict, peer_info):
                 commands.append("show ip bgp neighbors %s vrf all" % neighbor_address)
                 summary_commands.append("show ip bgp summary vrf all")
             elif peer_ver == 6:
-                commands.append("show ipv6 bgp neighbors %s vrf all" % neighbor_address)
+                commands.append("show ipv6 bgp %s vrf all" % neighbor_address)
```
[the remark this answers is not in the document]
Comment (emfusion) [reply napalm/eos/eos.py:1796]: no. this was an error did not mean to remove it

File: napalm/eos/eos.py (hunk 3 above)
Comment (emfusion) [reply napalm/eos/eos.py:1758]: the other devices share the naming so adding a separate counter for EOS will also mean adding for the other devices.

Comment (fdomain) [reply napalm/eos/eos.py:1758]: Ok fair enough.

File: napalm/eos/eos.py [hunk 5]
```diff
@@ -1682,13 +1682,26 @@ def _parse_per_peer_bgp_detail(peer_output):
...
                 "advertised_prefix_count",
                 "received_prefix_count",
+                "advertised_ipv6_prefix_count",
+                "received_ipv6_prefix_count",
             ]
 
             peer_details = []
 
+            # determine if in multi-agent mode to get correct extrator_type
+            commands = [
```
[the remark this answers is not in the document]
Comment (emfusion) [reply napalm/eos/eos.py:1692]: I did that at start but adding correct testing was a pain. 
If more discrepancies are found, this is something that could be done/added later. For now, I think this is the best way to do

File: napalm/eos/eos.py [hunk 6]
```diff
@@ -1699,7 +1712,7 @@ def _parse_per_peer_bgp_detail(peer_output):
                     True if item["local_address"] else False
                 )
                 item["multihop"] = (
-                    False if item["multihop"] == 0 or item["multihop"] == "" else True
+                    False if item["multihop"] == "0" or item["multihop"] == "" else True
```
Comment (mirceaulinic) [inline napalm/eos/eos.py:1715]: In the TexFSM template, the value would seem to be extracted as a number (`Value multihop (\d+)`). Could you clarify why comparing against string (`"0"`) here?

File: napalm/eos/eos.py [hunk 7]
```diff
@@ -1682,13 +1682,26 @@ def _parse_per_peer_bgp_detail(peer_output):
...
+                "received_ipv6_prefix_count",
             ]
 
             peer_details = []
 
+            # determine if in multi-agent mode to get correct extractor_type
+            commands = [
+                "show running-config | include service routing protocols model multi-agent"
+            ]
+            is_multi_agent = self.device.run_commands(commands, encoding="text")[0].get(
```
Comment (mirceaulinic) [inline napalm/eos/eos.py:1695]: Can probably get rid of the `commands` variable:

```suggestion
            is_multi_agent = self.device.run_commands(
                [
                    "show running-config | include service routing protocols model multi-agent"
                ],
                encoding="text",
            )[0].get()
```

**Review by mirceaulinic — APPROVED** [review]
Looks good overall, only had one question and a minor nit.

File: napalm/eos/eos.py (hunk 6 above)
Comment (emfusion) [reply napalm/eos/eos.py:1715]: Yes, so the TexFSM template is expecting a regular expression (\d+) for the multihop variable but the output remains a string : [example](https://pyneng.readthedocs.io/en/latest/book/21_textfsm/README.html#:~:text=The%20result%20of%20script%20execution)
that is also why the eos.py file still uses the "napalm.base.helpers.convert()" function to convert some of the extracted variables to int.

**Comment by bewing** [comment]
Little late the the party here, apologies.  

Are the two output formats so different we can't use a single template?  Recognizing the start of the next record should allow one to close out the previous record -- this approach is used in the VRF template: https://github.com/napalm-automation/napalm/blob/develop/napalm/eos/utils/textfsm_templates/vrf.tpl#L7

**Comment by mirceaulinic** [comment]
Merged as-is, but feel free to follow up with another PR and we shall review! 😄

**Comment by emfusion** [comment]
> @emfusion have you had a chance to look at Brandon's suggestion above? This still LGTM otherwise...

Hi, I unfortunately havent had the time to look into it (forgot how time consuming newborns are 😆). It's on my To Do list and will definitely jump into it soon

**Pull request merged**